### PR TITLE
Create ros2 implementation of basic_autonomy

### PR DESCRIPTION
**Pull request by JonSmet**
# PR Details
## Description

This PR creates a basic_autonomy_ros2 package, which is a ROS2 implementation of the basic_autonomy library. 

Related carma-utils PR for required basic_autonomy_ros2 dependencies: [Click Here](https://github.com/usdot-fhwa-stol/carma-utils/pull/119)

## Related Issue

Issue #1754 

## How Has This Been Tested?

Built locally and all unit tests pass.

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) 
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: basic_autonomy_ros2/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        src/basic_autonomy.cpp
+        src/helper_functions.cpp
+        src/log/log.cpp
+        src/smoothing/BSpline.cpp
+        src/smoothing/filters.cpp
+)
+
+target_link_libraries(
+        ${node_lib}
+        ${Boost_LIBRARIES}
```
Comment (msmcconnell) [inline basic_autonomy_ros2/CMakeLists.txt:47]: Is this line needed? Should there be find_package for Boost as well or adding it to the package.xml?

**Review by msmcconnell — CHANGES_REQUESTED** [review]

File: basic_autonomy_ros2/include/basic_autonomy_ros2/basic_autonomy.hpp [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+                    ((mvr).type == carma_planning_msgs::msg::Maneuver::LANE_CHANGE ? (mvr).lane_change_maneuver.property :\
+                        (mvr).type == carma_planning_msgs::msg::Maneuver::LANE_FOLLOWING ? (mvr).lane_following_maneuver.property :\
+                        (mvr).type == carma_planning_msgs::msg::Maneuver::STOP_AND_WAIT ? (mvr).lane_following_maneuver.property :\
+                        throw new std::invalid_argument("GET_MANEUVER_PROPERTY (property) called on maneuver with invalid type id"))))))
+
+namespace basic_autonomy
+{
+    namespace waypoint_generation
+    {
+        static const std::string BASIC_AUTONOMY_LOGGER = "basic_autonomy";
```
Comment (adev4a) [inline basic_autonomy_ros2/include/basic_autonomy_ros2/basic_autonomy.hpp:61]: Just to follow the same syntax as other packages, can we give define this as the complete namespace? basic_autonomy::waypoint_generation

File: basic_autonomy_ros2/src/basic_autonomy.cpp [hunk 3]
```diff
@@ -0,0 +1,1113 @@
...
+            }
+
+            std::vector<carma_planning_msgs::msg::TrajectoryPlanPoint> traj;
+            traj.reserve(points.size());
+
+            for (size_t i = 0; i < points.size(); i++)
+            {
+                carma_planning_msgs::msg::TrajectoryPlanPoint tpp;
+                rclcpp::Duration relative_time(times[i] * 1e9); // Conversion of times[i] from seconds to nanoseconds
+                tpp.target_time = startTime + relative_time;
```
Comment (adev4a) [inline basic_autonomy_ros2/src/basic_autonomy.cpp:720]: Curious about this one, does this compile? Can you directly assign rclcpp::Time to builtin_interfaces/Time

File: basic_autonomy_ros2/src/basic_autonomy.cpp [hunk 4]
```diff
@@ -0,0 +1,1113 @@
...
+                                                                                following_lanelet_centerline.end());
+                }
+            }
+
+            return centerline_points;
+        }
+
+       std::vector<std::vector<lanelet::BasicPoint2d>> resample_linestring_pair_to_same_size(std::vector<lanelet::BasicPoint2d>& line_1, std::vector<lanelet::BasicPoint2d>& line_2){
+            
+            rclcpp::Clock clock;
```
Comment (adev4a) [inline basic_autonomy_ros2/src/basic_autonomy.cpp:427]: This clock is never assigned a value. You can get the simulated time from this->now(), that should eliminate the need for creating a clock

Comment (JonSmet) [reply basic_autonomy_ros2/src/basic_autonomy.cpp]: Updated (and tested/verified) to use std::chrono as discussed.

**Pull request merged**